samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java

```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
+          "SELECT * FROM WorkList WHERE is_done is false and generated_value= '%s';";
+      Callable<Integer> callable =
+          new WorkItemProcessor<>(
+              workItemCount, i, NUM_PROCESSORS, dbClient, lockQueryInTransaction);
+      // Submitting each thread
+      Future<Integer> future = executorService.submit(callable);
+      futures.add(future);
+    }
+    int totalProcessed = 0;
+    // Calling the get on the future to wait for thread to finish and get the result.
+    try {
+      for (Future<Integer> future : futures) {
+
+        totalProcessed += future.get();
+      }
+
+      executorService.shutdown();
+      // Waiting for all thread to finish
+      executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public void doWorkSingleTransactionParallelNonLocking(DatabaseClient dbClient) {
+    // getting the available processor
+    ExecutorService executorService = Executors.newFixedThreadPool(NUM_PROCESSORS);
+
+    List<Future<Integer>> futures = new ArrayList<>();
+    // This is the store the count of total number of workItem(row) which has not yet completed
+    // means  is_done is false.
+    List<String> workItemCount = new ArrayList<>();
+    // Below logic is reading the id(primary key) for the rows which are having is_done=false. This
+    // will help us to read the different ids by different thread randomly.
+    try (ResultSet resultSet =
+        dbClient
+            .singleUse() // Execute a single read or query against Cloud Spanner.
+            .executeQuery(Statement.of("SELECT id FROM WorkList WHERE is_done is false;"))) {
+      while (resultSet.next()) {
+        workItemCount.add(resultSet.getString(0));
+      }
+    }
+
+    if (workItemCount.size() == 0) {
+      System.out.println(
+          "There is no workItem left to process, please update the is_done value to false, or add more work item in the table");
+      return;
+    }
+
+    // Assigning each thread the amount of work equally
+    for (int i = 0; i < NUM_PROCESSORS; ++i) {
+      // Each thread get the index from this loop, total workItem to be processed and number of
+      // thread (numProcessors) available for processing.
+      // E.g In case of 10 threads and 30 workItem the assignment for this method would be like
+
+      // Thread 1: ThreadIndex(i) 0, workitem 30
+      // Thread 2: ThreadIndex(i) 1  workitem 30
+      // Thread 3: ThreadIndex(i) 2  workitem 30
+      // ...
+      // Thread 10: ThreadIndex(i) 9  workitem 30
+      // These values will be determined by received by each thread(an object of WorkItemProcessor)
+      // while executing based on the input values passed here.
+      String nonLockQueryInTransaction =
+          "SELECT * FROM WorkList WHERE is_done is false and id= '%s';";
+      Callable<Integer> callable =
+          new WorkItemProcessor<>(
```

Comment:
This is a lot of indirection. I would prefer that we write these as one function and get rid of the WorkItemProcessor class.